Added global namespacing helpers to fix issue in dimensiondata module

tonybaloney opened the pull request:
### Fix Dimension Data Cloud module
### What issues does this PR fix or reference?
### Previous Behavior

PyLint raise a series of global methods that were not used, I removed them, https://github.com/saltstack/salt/commit/0ca43ba6b80c30b6a98dd17e52217b4bdf22e86d#diff-3dcc56f6a9e83d8023c42a7608157604
### Tests written?

No

Commit:
allow for specifying ssh_args to SSHConnection

Commit:
initial commit of nxos proxy minion

cmercier committed:
firewalld state/module : add runtime and persist + add rich_rules + merge bind and present

skizunov committed:
PY3 support for Windows master/minion using TCP transport

This allows masters and minions to run on Windows using Python 3. Tested
with Python 3.5.1 64-bit on Windows 7. Only the TCP transport has been
ported. It will likely not work with other transports such as ZeroMQ or
RAET without further work.

Since there were many changes necessary to make this work, here is a list
of common changes:
- Python 3 distinguishes `bytes` from `str`. In some cases `bytes` must be
used (eg read/write to binary file). In other cases `str` must be used
(eg general Salt dictionary look-ups, read/write to text file). Due to
this, there are a lot of extra uses of `salt.utils.to_bytes` in necessary
locations.
- Use `six.itervalues(varname)` instead of `varname.itervalues()`.
- In Python 2.6 and 2.7, the 'b' prefix is ignored. So instead of

```
if six.PY2:
    some_var = 'some_value'
else:
    some_var = b'some_value'
```

we choose to simply use:

```
some_var = b'some_value'

```

`salt/engine/__init__.py`:
- In Python 3, `engine.items()` will return an iterator instead of
a list. Due to this, needed to change `engine.items()[0]` to
`list(engine.items())[0]`.

`salt/minion.py`:
- `contextlib.nested` no longer supported in Python 3. Use
`contextlib.ExitStack` in this case.
- In `Minion.handle_event`, in PY3, `package` is of type `bytes` and `tag`
is of type `str`. Due to this, do all string searches via `tag` instead
of `package` since you can't search for a `str` in a `bytes` object.

`salt/payload.py`:
- Added `encoding` parameter to `Serial.loads`. See description in comments
for details.
- Added `use_bin_type` parameter to `Serial.dumps`. See description in
comments for details.
- When reading / writing local files, encode using `use_bin_type=True` and
decode using `encoding='utf-8'` to distinguish `bytes` and `str` types.

`salt/transport/frame.py`:
- Added `frame_msg_ipc`. This is used for IPC communications where it is
safe to break the wire protocol and so we encode using `use_bin_type=True`.
- Added `decode_embedded_strs`. This is used when it is not safe to break
the wire protocol (eg external communications). This will convert `bytes`
objects to `str` objects if possible. This will search for such objects
within dicts and lists.

`salt/transport/ipc.py`:
- Use `salt.transport.frame.frame_msg_ipc` to encode using
`use_bin_type=True`.
- Use `encoding='utf-8'` to decode such messages and ensure proper
distinction of `bytes` and `str` types.

`salt/transport/tpc.py`:
- Since we need to preserve the wire protocol, we don't use
`use_bin_type=True`. When decoding, we use `decode_embedded_strs` to
convert the embedded `bytes` objects that can be converted to `str`.
- `urlparse` doesn't exist in PY3. Use `urllib.parse` instead.
- `sys.maxint` not supported in PY3. Use `int((1 << 31) - 1)` instead.
Hence `sys.maxint - 1` becomes `int((1 << 31) - 2)`.

`salt/utils/__init__.py`:
- `libc = ctypes.cdll.LoadLibrary(ctypes.util.find_library("c"))` returns
`TypeError` in PY3. This is because `ctypes.util.find_library("c")`
returns `None` and in PY3 `ctypes.cdll.LoadLibrary(None)` is a
`TypeError`.
- On Windows, `salt.utils.fopen` would always read and write files in
binary mode. This has been changed for Python 3 since `bytes` must be
used for binary mode and `str` must be used for text mode and forcing
binary mode would screw things up if you want to read/write to `str`
objects. Preserved original behavior on PY2.
- For `salt.utils.fopen`, when reading and writing text files in PY3, if
an encoding is not specified, it will choose 'utf-8'.

`salt/utils/args.py`:
- In Python 3.5, every time `inspect.getargsspec` is used, a warning would
appear announcing its deprecation (it will be removed entirely in Python
3.6). Due to this, implemented our own version using
`inspect.getfullargspec`.

`salt/utils/event.py`:
- Since all functionality here is used only in IPC (the mechanism for
firing an event from minion to master is outside this file), we can break
the wire protocol and hence encode using `use_bin_type=True` and decode
using `encoding='utf-8'`.

`salt/utils/vt.py`:
- `_subprocess` is not available in PY3 under Windows, so we use an
alternate method to invoke `TerminateProcess` and `GetExitCodeProcess`.

`setup.py`:
- In PY3, `req.read(4096)` returns a `bytes` object. If you try to for
loop through it, each element is an `int` which represents an individual
byte. Trying to write an `int` to `wfh.write` raises an exception. Due to
this, use an alternate approach when writing from a
`http.client.HTTPResponse` object to file in PY3.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

mirceaulinic committed:
Tiny doc fix

mirceaulinic committed:
Ready to go

Commit:
Fix test failures

Commit:
Merge pull request #32119 from mirceaulinic/CF-DOCFIX

Tiny documentation fixes for napalm_ntp module

markuslang79 committed:
switched to InfluxDB 0.9+ without using tags or changing the existing queries

Commit:
Merge pull request #32107 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

Commit:
Fixes the nova cloud driver so it can parse OpenStack ip addresses.

Added code after the main ip address parsing that, if no ip addresses
are found, will attempt to parse the data structure as provided by
OpenStack (Kilo). The existing parsing, which works for Rackspace
is not affected.

Commit:
Merge pull request #32110 from rallytime/mac-desktop-tests

Add mac_desktop integration tests

Commit:
Merge pull request #32097 from blarghmatey/elasticsearch_event_return

Added `event_return` to Elasticsearch returner

Commit:
Merge pull request #31859 from hoonetorg/feature/module_state_pcs

initial commit for a pcs module/state

Commit:
Lint

Commit:
Merge pull request #32155 from cachedout/lint_32008

Lint 32008

Commit:
Add quotes to cron doc

Closes #31953

Commit:
Merge pull request #32157 from cachedout/issue_31953

Add quotes to cron doc

Commit:
Try/catch junos import

Commit:
Merge pull request #32159 from cachedout/fixup_31817

Try/catch junos import

abednarik committed:
states.boto_rds: __virtual__ return err msg.

Added call to missing_fun_string() in order to display module return message.

Fixes #32123.

messense committed:
Fix supervisord state bug when name contains `:*`

Commit:
Merge pull request #32172 from messense/fix-supervisord

Fix supervisord state bug when name contains `:*`

Commit:
Merge pull request #32161 from abednarik/virtual_ret_boto_rds_state

states.boto_rds: __virtual__ return err msg.

Commit:
Merge pull request #32138 from tegbert/fix-nova-driver-bug-parsing-ip-address

Fix nova driver bug parsing ip address

Commit:
Merge branch '2016.3' into 'develop'

Conflicts:
  - doc/topics/releases/2015.5.10.rst
  - doc/topics/releases/2015.8.8.rst

dongweiming committed:
Fix psutil.cpu_times unpack error

Commit:
Lint develop branch

Commit:
Lint develop branch

Commit:
Pylint fix

Commit:
Merge pull request #32180 from rallytime/lint-develop

Lint develop branch

Commit:
Merge pull request #32179 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

Commit:
Merge pull request #32182 from dongweiming/fix-ps

Fix psutil.cpu_times unpack error

Commit:
Implement a feature for module/state module pcs to be able to do changes in a CIB-file rather then manipulating the live-CIB.

This will be required for resources where commiting changes to the live-CIB atomically will destroy the cluster.

Example:
Master/Slave resources (like DRBD f.e.):
1. create a standard resource(ocf:linbit:drbd f.e.)
2. create a master slave set of the standard resource

These 2 changes should be commited at once to the live-CIB.

This can be done
- creating a CIB-file from the live-CIB
- adding the 2 resources in the CIB-file
- pushing the changed CIB-file to the cluster as the new live-CIB

With this patch it is possible to do this in a formula like this:

```
 # create a CIB-file from live-CIB (/some_path/<cibname>)
 # if it already exists and has the same content as the live-CIB
 # nothing is done
pcs_stonith__cib_created_cib_for_stonith:
  pcs.cib_created:
    - cibname: cib_for_stonith

 # create 2 resources in the CIB-file
pcs_stonith__create_my_fence_eps:
  pcs.stonith_created:
    - stonith_id: my_fence_eps
    - stonith_device_type: fence_eps
    - stonith_device_options:
        - 'pcmk_host_map=node1.example.org:01;node2.example.org:02'
        - 'ipaddr=myepsdevice.example.org'
        - 'power_wait=5'
        - 'verbose=1'
        - 'debug=/var/log/pcsd/my_fence_eps.log'
        - 'login=hidden'
        - 'passwd=hoonetorg'
    - cibname: cib_for_stonith

pcs_stonith__create_my_fence_apc:
  pcs.stonith_created:
    - stonith_id: my_fence_apc
    - stonith_device_type: fence_apc
    - stonith_device_options:
        - 'pcmk_host_map=node1.example.org:01;node2.example.org:02'
        - 'ipaddr=myapcdevice.example.org'
        - 'power_wait=5'
        - 'verbose=1'
        - 'debug=/var/log/pcsd/my_fence_apc.log'
        - 'login=hidden'
        - 'passwd=hoonetorg'
    - cibname: cib_for_stonith

 # commit the 2 created resources at once to the live-CIB from the CIB-file
 # if the CIB-file did not change since the creation with
 #   pcs.cib_created: pcs_stonith__cib_created_cib_for_stonith
 # then push nothing
 # (obviously the 2 resources above were already in the CIB-file and therefor
 # in the live-CIB, so the live-CIB should already contain the 2 resources above)
pcs_stonith__cib_pushed_cib_for_stonith:
  pcs.cib_pushed:
    - cibname: cib_for_stonith
```

Commit:
Merge pull request #32181 from jtand/fix_test_kill

Fix test kill

castor-morveer committed:
adding execution module for AWS CodeCommit, no tests yet

castor-morveer committed:
adding execution module for AWS CodeCommit, no tests yet

Commit:
service.py documentation update for 32084

Commit:
documentation-fix-31931

castor-morveer committed:
added state module for AWS CodeCommit, still no tests

castor-morveer committed:
added state module for AWS CodeCommit, still no tests

Commit:
documentation fix issue 31844

Commit:
Merge pull request #32197 from twellspring/documentation-fix-issue-31844

documentation fix issue 31844

Commit:
Merge pull request #32185 from hoonetorg/feature/pcs_use_cib

pcs: implement commiting resources to a CIB-file rather than the live-CIB and push this file then to the live-CIB

Commit:
add support for other types of projections in global secondary indexes

Commit:
Feature: sdb.delete

Adds sdb.delete operation support to execution modules, runners and
utils. Docs updated as well.

m03 committed:
Reworked function args post review

Commit:
Whitespace fix.

Commit:
Whitespace fix.

Commit:
Merge pull request #32208 from rallytime/merge-32192

Merge #32192 with pylint fix

Commit:
Merge pull request #32210 from rallytime/merge-32193

Merge #32193 with pylint fix

m03 committed:
Revise docstrings

Commit:
add nxos state

Commit:
fix up roles for nxos proxy minion user

Commit:
add removing users

Commit:
remove extra return line

Commit:
Allow for specifying not using the random module

The Random module can lead to non alpha numeric characters in the
password.  Since we also use this for creating the salt, some devices
cannot have these extra characters in their salts, so allow to turn this
feature off as needed.

Commit:
use encrypted pass everywhere for nxos

Commit:
Fixed boto_lambda_test

VpcConfig was not added to function_ret, which was causing
test_that_when_describing_function_it_returns_the_dict_of_properties_returns_true
to fail.

Commit:
clean up sets

Commit:
fix regex for plugins

Commit:
allow accepting the ssh key

Commit:
log error if multiprocessing is True

nxos uses the sshconnection, which has to have everything live inside
the main process where the ssh connection is created, otherwise the
connection will appear to be closed.

ticosax committed:
Improve error message in case of API mismatch

fixes #31983

markuslang79 committed:
query http api to import the matching influxdb client version

ticosax committed:
only 2 argument are passed to this template.

markuslang79 committed:
fixed version check via the /ping HTTP API endpoint

Commit:
Merge pull request #32238 from ticosax/fix-gitfs-error-message

[gitfs] only 2 argument are passed to this template when render error message

Commit:
Merge pull request #32228 from jtand/boto_lambda_test_fix

Fixed boto_lambda_test

Commit:
Merge pull request #32232 from ticosax/improve-error-message

[dockerng ]Improve error message in case of API mismatch

Commit:
Bugfix: match.filter_by minion_id regression fix

There is a regression since e61ac75d6f04e694bdb6440fffe5634fcbba7c8b.
Even when minion_id is provided it tries to match against id from opts.

Commit:
Merge pull request #32202 from lyft/multiple_projection_types_in_GSI

add support for other types of projections in global secondary indexes

Commit:
Merge pull request #31865 from m03/add-win_smtp_server

Add execution and state modules for managing Windows SMTP server

markuslang79 committed:
cached result of the version check via the /ping HTTP API endpoint

Commit:
Merge pull request #32241 from Ashald/bugfix/match-filter-by-minion-id

Bugfix: match.filter_by minion_id regression fix

Commit:
Make return types consistent in sdb_set

castor-morveer committed:
updated template mapping to adopt recent AWS changes in stage variables mapping

castor-morveer committed:
updated template mapping to adopt recent AWS changes in stage variables mapping

castor-morveer committed:
updated template mapping to adopt recent AWS changes in stage variables mapping

castor-morveer committed:
updated template mapping to adopt recent AWS changes in stage variables mapping

Commit:
Revert "added state module for AWS CodeCommit, still no tests"

This reverts commit fbce562b7002b4e6e963f4f24b8de782385e1917.

Commit:
Revert "added state module for AWS CodeCommit, still no tests"

This reverts commit fbce562b7002b4e6e963f4f24b8de782385e1917.

Commit:
Revert "adding execution module for AWS CodeCommit, no tests yet"

This reverts commit 43e7fb1bdef92644978190da573dd7f2ac798b39.

Commit:
Revert "adding execution module for AWS CodeCommit, no tests yet"

This reverts commit 43e7fb1bdef92644978190da573dd7f2ac798b39.

Commit:
Merge pull request #32247 from Ashald/bugfix/utils-sdb-set-return

Make return types consistent in sdb_set

SeverinLeonhardt committed:
Fix check_exists in alternatives module

The output of update-alternatives --display contains not only a path but
a path followed by its priority. The check must therefore use startswith
on each output line.

Commit:
mysql: use replication safe create db statement

Commit:
Merge pull request #32266 from dropje86/mysql_safe_createdb

mysql: use replication safe create db statement

Commit:
Fix unit test for mysql

Refs #32266

Commit:
Merge pull request #32270 from cachedout/fixup_32266

Fix unit test for mysql

Commit:
Merge pull request #32242 from Ashald/feature/sdb-delete

Feature: sdb.delete

Commit:
add find and replace for nxos

Commit:
Merge pull request #32120 from cmercier/firewalld

firewalld state/module : add runtime and persist + add rich_rules + merge bind and present

Commit:
added $util.escapeJavaScript to mapped stageVariable values.

Commit:
added $util.escapeJavaScript to mapped stageVariable values.

Commit:
optional range_key

Commit:
Merge pull request #32282 from lyft/optional_range_key

Optional range_key

Commit:
add, delete, and replace configs in NXOS

Allow for adding configs, deleteing configs, or replacing partial or
full config lines in nxos.

Commit:
include grains for nxos

Commit:
add_permission call did not have the region, key, keyid and profile passed in, hence failing on creating a new lambda if permissions were included.

Commit:
Merge branch 'boto_lambda' into boto_apigateway

Commit:
add documenation for nxos switch proxy minion

Commit:
pylint fixes

Commit:
clean up proxy for nxos to be more stable

Commit:
allow for multiprocessing with nxos

Commit:
Merge pull request #32290 from kraney/boto_lambda

add_permission call for a new lambda function present state goes to wrong aws region

Commit:
add warning for proxy_merge_grains_in_module

Commit:
Merge branch '2016.3' into 'develop'

Conflicts:
  - salt/modules/rh_service.py

Commit:
pylint fixes part 2

Commit:
Merge pull request #32296 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

nmadhok committed:
Fail when neither image nor clonefrom are specified for VMware cloud driver

tamentis committed:
Avoid pkg_add to go in interactive mode.

m03 committed:
Address win_iis issues 32049 and 32050

m03 committed:
Make docstrings consistent

carlpett committed:
Fix wrong exception type

Commit:
Merge branch 'develop' into boto_apigateway

Commit:
Merge pull request #32328 from carlpett/patch-2

Fix wrong exception type

Commit:
Merge pull request #32317 from m03/fix-32049-and-32050

Address issues #32049 and #32050, regarding win_iis

Commit:
Merge pull request #32313 from tamentis/develop

Avoid pkg_add to go in interactive mode.

Commit:
Merge pull request #32309 from nmadhok/patch-2

Fail when neither image nor clonefrom are specified for VMware cloud driver

Commit:
Merge pull request #32292 from gtmanfred/develop

Add proxy minion for Cisco NX OS Switches

carlpett committed:
Change order of precedence in opts when executing minion_runner

Commit:
Deleted win_servermanger_test.py

Commit:
Merge pull request #32334 from jtand/delete_win_servermanager_test

Deleted win_servermanger_test.py

Commit:
Merge branch 'boto_apigateway' of https://github.com/kraney/salt into boto_apigateway

Commit:
Merge pull request #32335 from carlpett/patch-3

Change order of precedence in opts when executing minion_runner

skizunov committed:
Use `next(iter(engine.items()))`

Use `next(iter(engine.items()))` instead of `list(engine.items())[0]`

nmadhok committed:
Check if profile key exists in vm_ dict

nmadhok committed:
Default to True and False instead of None

Ref PR #32309

jirikotlin committed:
Better documentation, host interface functions, modified states

markuslang79 committed:
fixed trailing whitespaces

SeverinLeonhardt committed:
Add missing closing parenthesis

Commit:
outputter virt_list does not exist anymore

naegelin committed:
Support Private Networking and Hostnames

What does this PR do?

Support passing in hostname so the spun up instance does not default to 'vultr' as hostname
Support enabling private networking in the salt cloud profile
Extend salt exceptions
Extend documentation to be more inclusive of vultr specific cloud profile
Previous Behavior

Hostnames were defaulted to 'vultr'
Non support for private networking

Tests written?

No - tested manually

Commit:
Merge pull request #32346 from nmadhok/patch-2

Default to True and False instead of None

Commit:
Merge pull request #32344 from nmadhok/patch-3

Check if profile key exists in vm_ dict

Commit:
Merge pull request #32329 from skizunov/develop2

PY3 support for Windows master/minion using TCP transport

Commit:
Merge pull request #32358 from arthurlogilab/patch-18

outputter virt_list does not exist anymore

Commit:
Merge pull request #32263 from SeverinLeonhardt/fix_alternatives.check_exists

Fix check_exists in alternatives module

Commit:
Merge pull request #32125 from pret79/develop

Added InfluxDB 0.9+ support to the influxdb_return.py class

naegelin committed:
Support Private Networking and Hostnames

naegelin committed:
Formatting fixes

naegelin committed:
Trailing whitespace....

Surely Jenkins can be configured to auto-correct this...

Commit:
error comments in some present state function reported incorrectly to result['common'] instead of result['comment']

Commit:
Merge pull request #32362 from naegelin/develop

Support Private Networking and Hostnames

Commit:
Merge branch '2016.3' into 'develop'

Conflicts:
  - pkg/osx/build.sh

abednarik committed:
Cleanup verify error messages.

Thi PR is a followup of @cachedout comments in #32321.

rajvidhimar committed:
minor bug fixes in call_rpc

Commit:
Merge pull request #32370 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

Commit:
Merge pull request #32387 from rajvidhimar/junos-call_rpc-changes

minor bug fixes in call_rpc

Commit:
Merge pull request #32386 from abednarik/abednarik_verify_msg_cleanup

Cleanup verify error messages.

Commit:
Merge pull request #32350 from jirikotlin/develop

Better documentation, host interface functions, modified states

Commit:
Merge pull request #32340 from kraney/boto_apigateway

updated request mapping template to properly map stageVariables

garethgreenaway committed:
Slack engine fix results

* Changing the way results from the Slack engine are displayed in Slack, rather than dump the results into the channel as a message we upload as a file.  This way Slack handles displaying as much of the content as it can and the full results are still available.

* Change type of file to be yaml so Slack will format it nicely.

* removing code in master.py

* engine configuration is now a dictionary of dicionaries, updating docstring for slack engine

Commit:
Change some config defaults in the Carbon release

Commit:
Merge pull request #32395 from kraney/boto_apigateway

error comments in some cases do not show up after a boto_apigateway present state

Jonher937 committed:
Update win_service.py

Space at the wrong place

Jonher937 committed:
Update win_service.py

One more space found

Commit:
Merge pull request #32402 from Jonher937/patch-1

Typo in win_service.py

Commit:
Merge pull request #32398 from rallytime/config-default-change

Change some config defaults in the Carbon release

toddtomkinson committed:
remove "duplicate" fetch array when updating a marathon app

fixes mesosphere/marathon#3054 for the marathon module by implementing
the same fix used in the marathon ui

Commit:
Merge pull request #32408 from toddtomkinson/fix_fetch_uri_mismatch

remove "duplicate" fetch array when updating a marathon app

m03 committed:
Add vdir functionality to win_iis (#32403)

moises-silva committed:
Improve Key.gen_keys() method to be more flexible and have better defaults (#32380)

* Allow passing extra keyword arguments to tweak the key generation

  process through the API without requiring special options to be

  present in the minion configuration options.



* Default the key generation directory to pki_dir when gen_keys_dir

  is not provided rather than failing with a KeyError exception.



* Default the keyname to 'minion' if no gen_keys option was provided.



* Return the fingerprint of the generated pub key.



Generating keys for a minion programatically with the default minion

configuration can now be done simply doing:



from salt.config import minion_config

from salt.key import Key



Key(minion_config(None)).gen_keys()



This will return the generated key fingerprint.



Note that the implementation for RaetKey has a significantly different

interface despite its inherited behavior from the Key class. An effort

was made to make the interface as close as possible, but, it does

not generate a PKI key pair for example, and no fingerprint is

returned when calling gen_keys().

skizunov committed:
TCP Load balancer for Windows Python 3 (#32401)

This will add load balancer functionality to the salt-master

when running on Windows Python 3. Unfortunately, on Python 2, the

limit for `worker_threads` remains 1 due to a lack of needed

functionality in Python 2. See issue #27367 for details.



salt/master.py:

- Modified the `tcp_only` check to only issue the warning if we are

using Windows Python 2 and `worker_threads` is not already set to 1.



salt/transport/tcp.py:

- Added global `USE_LOAD_BALANCER` to tell if we need to add the

extra load balancer functionality. Currently, it will be set if

we are using Windows and Python 3.

- Added `LoadBalancerServer`. This runs in its own process and will

listen for incoming connections. Each incoming connection will be

sent via multiprocessing queue to the workers. Since the queue is shared

amongst workers, only one worker will handle a given connection.

- Added `LoadBalancerWorker`. This inherits from `SaltMessageServer`

and is used in its place for `TCPReqServerChannel.req_server` when

`USE_LOAD_BALANCER` is True. It adds the additional functionality of

getting connections from the shared multiprocessing queue that

`LoadBalancerServer` sends connections to and processes those

connections.



Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov committed:
PY3: Fix `trim_dict` to work with `use_bin_type` (#32406)

salt/utils/dicttrim.py:

- Changed `salt.utils.dicttrim.trim_dict` to take an extra parameter

`use_bin_type`. Hence, when handling this data, it will decode

using `encoding='utf-8'` and encode using `use_bin_type=True`.



salt/utils/event.py:

- Since events use `use_bin_type=True` for Python 3, when calling

`salt.utils.dicttrim.trim_dict` set `use_bin_type=six.PY3`



salt/minion.py:

- Minion._process_beacons also uses `salt.utils.dicttrim.trim_dict`.

However, it doesn't seem to be referenced anywhere, so it has been

removed entirely. It wasn't PY3 compliant for other reasons anyway.

Surprisingly, `salt.utils.dicttrim` wasn't even imported in

minion.py to begin with.



Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

hemebond committed:
Discard duplicate YAML nodes when merging anchors to avoid duplicate … (#32042)

* Discard duplicate YAML nodes when merging anchors to avoid duplicate node errors. Add yamlloader unit tests.

Fixed #14550 Using YAML anchors/references in Pillars causes conflicting IDs

* Fix lint errors. Make external function a static method.

* Fix remaining lint errors in unit test file.

eradman committed:
Do not assume temporary files have the user's gid

On BSD-like systems (including MacOS) files created in /tmp (or any dir
with mod 1777) do not have the user's group. Therefore the st_gid for
files created in /tmp will usually be wheel. Easiest thing to do is to
skip this test.

eradman committed:
Unbreak timezone test for Gentoo (#32420)

Gentoo and Debian both use use /etc/timezone

See https://wiki.gentoo.org/wiki/System_time

eradman committed:
Unbreak cron.file (#32396)

* Unbreak cron.file

An extra argument (skip_verify) was recently added to get_managed(),
resulting in this error:

  Unable to manage file: get_managed() takes exactly 11 arguments (10 given)

Also eliminate a variable that can remain uninitialized when running
`salt-call`

* Correct pylint error

Commit:
Fix issue that #32420 exposed in timezone_test

Commit:
Merge branch '2016.3' into 'develop'

Conflicts:
  - salt/cli/daemons.py
  - salt/modules/win_servermanager.py
  - tests/unit/modules/win_servermanager_test.py

Commit:
Merge pull request #32428 from jtand/timezone_test_fix

Fix issue that #32420 exposed in timezone_test

Commit:
Merge pull request #32422 from eradman/env_test

Do not assume temporary files have the user's gid

eradman committed:
Skip PAM test on OpenBSD (#32439)

Will always fail since OpenBSD does not have have PAM

Commit:
Removed useless slack_notify unit tests

Commit:
Merge pull request #32431 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

eradman committed:
Use reload_module from six for compatibility with Python3 (#32438)

reload() was moved to importlib in 3.4

kaihowl committed:
Fix broken links in topics/development/contributing.rst

Commit:
Merge pull request #32461 from kaihowl/fix-links

Fix broken links in topics/development/contributing.rst

suhasgaddam committed:
Add marathon restart functon to state and exec modules (#32470)

* Add marathon state and module to restart app

* Add a restart=False default for marathon restart

* Update state name to marathon_app.running

Update returned comment based on restart option. Add check for exception
from salt.utils.http.query.

* Update documentation on state and exec module function

* Cleanup code block examples in marathon exec module

hoonetorg committed:
PCS state module: feature set for creating clusters and resources (#32468)

With this modules/state module it's now possible to do a
- cluster node authorization
- cluster setup
- setting cluster properties, resource defaults and resource op defaults.
- creating fence devices
- creating resources and constraints
- doing multiple modifications to the CIB in a cibfile first and
- pushing all the changes done to a cibfile at once to the live CIB of the cluster

See the walkthrough in the doc section of salt/states/pcs.py.

m03 committed:
Add sourcepath check to win_iis.create_vdir (#32442)

m03 committed:
Add applications to win_iis (#32483)

Commit:
[develop] Merge forward from 2016.3 to develop (#32494)

* fix sorting by latest version when called with an attribute

* remove reference to master_alive_check

* Fixes saltstack/salt#28262

* Resolve memory leak in authentication

* outputter virt_list does not exist anymore

* Update proxmox documentation

* Fix documentation on boto_asg and boto_elb modules and states

* modules.win_timezone: don't list all zones in debug log

* Correcty index glusterfs bricks

Fixes issue #32311

* Cleaner deprecation process with decorators

* Add deprecation decorator scaffold

* Capture type error and unhandled exceptions while function calls

* Aware of the current and future version of deprecation

* Implement initially is_deprecated decorator

* Add an alias for the capitalization

* Fix capitalization easier way

* Remove an extra line

* Add successor name to the deprecation decorator.

* Granulate logging and error messages.

* Implement function swapper

* Raise later the caught exception

* Clarify exception message

* Save function original name

* Remove an extra line

* Hide an alternative hidden function name in the error message, preserving the error itself

* Rename variable as private

* Add a method to detect if a function is using its previous version

* Message to the log and/or raise an exception accordingly to the status of used function

* Log an error along with the exception

* Add internal method documentation

* Add documentation and usage process for decorator "is_deprecated"

* Add documentation and process usage for the decorator "with_deprecated"

* Hide private method name

* Fix PEP8, re-word the error message

* Deprecate basic uptime function

* Add initial decorator unit test

* Rename old/new functions, mock versions

* Move frequent data to the test setup

* Add logging on EOL exception

* Rename and document high to low version test on is_deprecated

* Implement a test on low to high version of is_deprecated decorator

* Add a correction to the test description

* Remove a dead code

* Implement a test for high to low version on is_deprecated, using with_successor param

* Correct typso adn mistaeks

* Implement high to low version with successor param on is_deprecated

* Setup a virtual name for the module

* Implement test for with_deprecated should raise an exception if same deprecated function not found

* Implement test for with_deprecated an old function is picked up if configured

* Correct test description purpose

* Implement test with_deprecated when no deprecation is requested

* Add logging test to the configured deprecation request

* Add logging testing when deprecated version wasn't requested

* Implement test EOL for with_deprecated decorator

* Correct test explanation

* Rename the test

* Implement with_deprecated no EOL, deprecated other function name

* Implement with_deprecated, deprecated other function name, EOL reached

* Add test description for the with_deprecated + with_name + EOL

* Fix confusing test names

* Add logging test to the is_deprecated decorator when function as not found.

* Add more test point to each test, remove empty lines

* Bugfix: at certain conditions a wrong alias name is reported to the log

* Fix a typo in a comment

* Add test for the logging

* Disable a pylint: None will _never_ be raised

* Fix test for the deprecated "status.uptime" version

* Bugfix: Do not yank raised exceptions

* Remove unnecessary decorator

* Add test for the new uptime

* Add test for the new uptime fails when /proc/uptime does not exists

* Rename old test case

* Skip test for the UTC time, unless freeze time is used.

* Fix pylint

* Fix documentation

* Bugfix: proxy-pass the docstring of the decorated function

* Lint fix

* Fixes saltstack/salt#28262 for 2015.5 branch

* Update master config docs

* Improve git_pillar documentation/logging

* Add note about different behavior of top file in git_pillar

* Make log entry for a missing pillar SLS file more accurate for git_pillar

* FreeBSD supports packages in format java/openjdk7 so the prior commit broke that functionality. Check freebsd/pkg#1409 for more info.

* FreeBSD supports packages in format java/openjdk7 so the prior commit broke that functionality. Check freebsd/pkg#1409 for more info.

* Update glusterfs_test to be inline with #32312

* Fix salt-cloud paralell provisioning

Closes #31632

* Ignore Raspbian in service.py __virtual__ (#32421)

* Ignore Raspbian in service.py __virtual__

This prevents more than one execution module from trying to load as the
service virtual module.

Refs: #32413

* pack __salt__ before loading provider overrides

We can (and should) pack here since we're just packing a reference to the
object. __salt__ needs to be available when we're loading our provider
overrides

* Fix broken __salt__ dict in provider override

Using ret.items() here sets ``__salt__`` to its items (tuple containing
function name and reference), breaking usage of ``__salt__`` inside
overridden functions.

* Merge #32293 with test fixes (#32418)

* Fix issue #11497

* Remove check for working directory presence in tests

* Fix Domainname introspection

Default value needs to be extracted from the container itself,
because dockerd set Domainname value when network_mode=host.

* Add pgjsonb_queue to queue doc index

* Pylint fixes

* Pass parser options into batch mode

Resolves #31738

* Changed the target file in file.symlink test (#32443)

* Argument name in docs should match actual arg name (#32445)

Fixes #31851

* tests.integration: bypass MacOS TMPDIR, gettempdir (#32447)

Updates 0edd532, 8f558a5.

When logging in as root over `ssh root@host`, `$TMPDIR` and
`tempfile.gettempdir()` are both set to a variation of:
```
/private/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/
```
When logging in as root over `sudo -i`, `$TMPDIR` is unset and
`tempfile.gettempdir()` is set to `/tmp`.

My guess is that the second case is an unintended or uncorrected omision
by Apple as they have introduced the longer, randomized temp path in a
recent version of MacOS.

* Issue #28706: Fix state user.present behavior. (#32448)

- As mentionned in issue #28706, state user.present no longer remove
      user from groups if the keyword 'groups' with empty value '[]' is not
      explicitly set, salt will assume current groups are still wanted.

* tests.integration: fix 4230c8a

* Move the tables of virtual modules to individual documentation pages

* Add new doc pages to toctree

* Add external ref to windows package manager docs

* Improve docstrings

* Add documentation on virtual module provider overrides to the module docs

* Clarify the scope of the provider param in states.

* Add link to provider override docs to all package providers

* Add link to provider override docs to all service providers

* Add link to provider override docs to all user providers

* dd link to provider override docs to all shadow providers

* Add link to provider override docs to all group providers

* Backport 31164 and 31364 (#32474)

* Don't send REQ while another one is waiting for response.

The message has to be removed from the queue the only *after* it's
already processed to don't confuse send() functionality that expects
empty queue means: there's no active sendings.

* Fixed zeromq ReqMessageClient destroy

* Add link to provider override docs to opkg.py

This is a companion to https://github.com/saltstack/salt/pull/32458, but
this module was not added until the 2016.3 branch, so the documentation
is being updated there for this module.

* Add documentation for some master/minion configs (#32454)

Refs #32400

Adds docs for:

- cli_summary
- event_return_queue
- event_return_whitelist
- event_return_blacklist
- file_recv_max_size
- fileserver_followsymlinks
- fileserver_ignoresymlinks
- fileserver_limit_traversal

* Automatically detect correct MySQL password column for 5.7 and fix setting passwords (#32440)

* Automatically detect MySQL password column

* Fix changing password in MySQL 5.7

* Fix lint test

* Fix unit tests (?)

They will still fail if "authentication_string" is legitimately the right column name, but I don't know what to do about that.

* Additional unit test fix

* Only unsub if we have a jid

Closes #32479

carsonoid committed:
Closes #31078 - Add options for syslog returner (#32464)

isntall committed:
Small typo fix (#32498)

Commit:
Add Cumulus support to aptpkg (#32504)

Cumulus is a Debian based switch OS, needs support for pkg installs.

ewalker11 committed:
Add ability to configure DynamoDB backups using pillars (#32503)

Commit:
Fixing Unicode Error / Accept Eula Windows Update Modules/States (#32502)

* Fixed issue #30980 (UnicodeEncodeError -> non ascii characters in update title)
added accept eula logic

* Fixed issue #30980 (UnicodeEncodeError -> non ascii characters in update title)
Fixed AcceptEula, didn't work correctly because of missing parentheses

tkwilliams committed:
Various boto2 module enhancements (#32496)

* ip_protocol.isdigit() fails when isinstance(ip_protocol, int)

* Fix """Warnings: 'name' is an invalid keyword argument for
'boto_iam.account_policy'. If you were trying to pass additional
data to be used in a template context, please populate 'context'
with 'key: value' pairs. Your approach will work until Salt Carbon
is out. Please update your state files."""

* Add boto_iam.get_all_users() module function

* Add boto_iam.get_all_groups() module function

* Update boto_iam.export_users() to use new get_all_users() function

* Add boto_iam.get_all_roles() module function

* Finally got around to implementing image_name option to boto_ec2.instance_present :)

* Add boto_secgroup.get_all_security_groups() function
Add options public_ip, allocation_id, and allocate_eip to boto_ec2.instance_exists()
Add option release_eip to boto_ec2.instance_absent()

* Boy, I REALLY disagree with PEP about the whole "whitespace after/before braces/brackets/parens" readability thing....

eradman committed:
Fix tests that assert CommandExecutionError (#32485)

Trying to assert that an exception was raised using
helper_open.write.assertRaises() is bogus--there is no such method. Use
standard unittest.assertRaises() instead.

thatch45 committed:
Fix rendering error with thorium docs

Commit:
Merge pull request #32516 from thatch45/thorium_doc

Fix rendering error with thorium docs

m03 committed:
Add site bindings to win_iis (#32530)

AndreiPashkin committed:
Document "grains" setting in the minion configuration reference (#32528)

Commit:
Fix salt-call on standalone minion case (#32521)

* skip connection error if master_type == disable

Signed-off-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>

* fix salt-call when master_type == disable

salt-call errors out on salt.pillar.get_pillar if master_type ==
disable. Skip get_pillar call on standalone minion case. Also skip code
that returns job info back to the master.

Signed-off-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>

abednarik committed:
Updated rsync module _virtual__. (#32524)

Small update to presnet a proper message when rsync command is not found and
module cannot be loaded.

AndreiPashkin committed:
Fix YAML-block in "Organizing Pillar data" part of "Salt Formulas" docs page (#32527)

Commit:
Add get_issue_comments function to the github module (#32522)

eradman committed:
Allow user login class to be empty on OpenBSD (#32518)

* Fix tests that assert CommandExecutionError (#32485)

Trying to assert that an exception was raised using
helper_open.write.assertRaises() is bogus--there is no such method. Use
standard unittest.assertRaises() instead.

* Allow user class to be blank

Attempts to address Issue#32512

eradman committed:
Avoid KeyError if cron change does not include a diff (#32513)

Prevents this exception:

    File "/tmp/saltenv/lib/python2.7/site-packages/salt/state.py", line 1703, in call
      **cdata['kwargs'])
    File "/tmp/saltenv/lib/python2.7/site-packages/salt/loader.py", line 1661, in wrapper
      return f(*args, **kwargs)
    File "/tmp/saltenv/lib/python2.7/site-packages/salt/states/cron.py", line 567, in file
      ret['changes'] = {'diff': ret['changes']['diff']}
    KeyError: 'diff'

As far as I can tell this line is a noop at best

twangboy committed:
Fix binary search and replace (#32542)

cro committed:
Expand on the open-source vs open-core FAQ

skizunov committed:
PY3: salt-key --gen-keys fix (#32546)

`__init__.py`:
- On certain invocations of salt-key (seen when invoked programmatically
on Windows without going through the command interpreter `cmd.exe`),
`sys.stdin` is None. This has only been seen on Python 3. So
`encoding = sys.stdin.encoding` raises an exception in this case.

`utils/__init__.py`:
- `pem_finger` reads a file in binary mode. So in Python 3, the return
type is bytes and not str. Simple fix by using `b''.join` instead of
`''.join`.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

thatch45 committed:
these are no longer loaded by the loader and proved to be (#32548)

more trouble that they were worth.

Commit:
Add html_url to "min" dictionary that is returned for issues (#32549)

deuscapturus committed:
Allow sqs messages to be deserialized from json (#32557)

This engine is most useful if users could deserialize json messages, making the data within usable in reactors.  For example, messages from the CloudWatch Event system.

cro committed:
Language clarification.

Commit:
Merge branch '2016.3' into 'develop'

Conflicts:
  - doc/ref/configuration/master.rst
  - salt/modules/glusterfs.py
  - salt/modules/service.py
  - tests/unit/modules/glusterfs_test.py

Commit:
Passing kwargs when creating mapper.

Commit:
Added import_keypair method for digital_ocean.

Commit:
Merge pull request #32566 from one-hero/cloud-client-action-kwargs

Passing kwargs when creating mapper.

Commit:
Merge pull request #32560 from rallytime/merge-develop

[develop] Merge forward from 2016.3 to develop

Commit:
Merge pull request #32547 from cro/opencorefaq

Expand on the open-source vs open-core FAQ

Commit:
Implemented code review requirements

gtmanfred committed:
make nxos module load init for multiprocess True (#32577)

This is required for test.ping to return True with `multiprocessing: True`
caused by the fact that the ssh connection has to be reconnected each
time the process is forked.

Commit:
Fixed lint failure in sqs_events (#32580)

Commit:
Better log message on minion restart if master couldn't be reached. (#32576)

felixhummel committed:
mention /etc/salt/minion_id in docs explicitly (#32573)

This file is essential if one needs to change a minion's ID. Showing the
full path to the file makes it easier searching for it, e.g. Google::

    http://blog.mindfab.net/2013/11/salt-changing-minions-hostname.html

Commit:
Merge pull request #32569 from one-hero/same-import-keypair

Added import_keypair method for DigitalOcean

Commit:
add nxos grains just like salt/grains/nxos.py

jacobhammons committed:
Refreshes the doc module list

jacobhammons committed:
Merge pull request #32595 from jacobhammons/mod-updates

Refreshes the doc module list

Commit:
strip extra whitespace from regex match on roles

Commit:
added WaitForRole and RoleRetries to update_function_config to cover for case of a newly created Role

jacobhammons committed:
This module is making a call that is breaking the doc build, removing until I can figure it out.

jacobhammons committed:
Merge pull request #32598 from jacobhammons/mod-updates

Docs for this module are breaking the build

Talkless committed:
Fix comments value in salt.states.pkgrepo example (#32604)

'comments' option adds '#' automatically. Example contains `#http://mirror.centos.org/centos/$releasever/os/$basearch/` string which becomes prefixed with '##' in generated file.

gtmanfred committed:
alphabetize directories for dynamic modules (#32599)

Also add engines and proxy minions to the list.

Commit:
Merge pull request #32597 from kraney/boto_lambda

added WaitForRole and RoleRetries to update_function_config

Commit:
Merge pull request #32594 from gtmanfred/develop

add nxos grains just like salt/grains/nxos.py

Commit:
Merge pull request #32596 from gtmanfred/fix_roles_nxos_user_state

strip extra whitespace from regex match on roles for nxos state

thatch45 committed:
Thormods (#32600)

* add basic runner caller to thorium

* add basic wheel module

* only start the reactor if it is configured and add it correctly to the engines

* fix lint and a syntax error I forget to get into the PR

* add break to speed things up a bit

* bah! whitespace!

Commit:
Feature minion.restart (#32593)

* Add minion.restart

* Improve minion.kill for use by minion.restart

  + Add "retcode" status
  + Wait until the process dies (add timeout value)
  + Provide meaningful "comment" information when appropriate

* Add minion.restart: kills minion with minion.kill and then starts a new minion process

  + Can provide a specific ``minion_restart_command`` arg vector in minion configuration
  + If no ``minion_restart_command`` is available then uses ``argv`` of current process
  + If minion is not daemonized then it will kill but not restart - allows for
    direct management of minion by init, systemd or other direct process
    manager.

* Add EX_TEMPFAIL to exitcodes for timeout situtations (used by minion.kill)

TODO: integration tests

* Appease the pylint gods.

* Tweak the returned data structure for minion.kill.

* Seed the ``minion.kill`` ``killed`` entry with ``None.``

quantonganh committed:
i32571-winrepo: `saltenv` is required to lookup the cache file (#32581)

* i32571-winrepo: `saltenv` is required to lookup the cache file

* `win_pkg` module: fix pep8 errors

* i32571-winrepo: purged/remove also need to know the `saltenv`

anlutro committed:
Fix some mistakes in the salt-ssh thin shell script (#32583)

* [[ is bash, not compatible with /bin/sh

* check if python command exists before calling it

The-Loeki committed:
Deprecate 'user' and 'group' in state cmd (#32613)

* Remove unused 'group' argument

* Fix unit testing of cmd.mod_run_check without group arg

* Deprecate 'user/group' in cmd.run

* Deprecate 'user'/'group' in cmd.script

* Deprecate 'user' in cmd.wait

* Deprecate 'user'/'group' in cmd.wait_script

* Fix mod_run_check without 'group'

* Push deprecation back one release

ksvasan committed:
Enable blocking on min_party until minimum set of nodes are available (#32592)

* Enable blocking on min_party until minimum set of nodes are available

* Fix the comment in min_party for salt-pr-lint-n

* Add test cases to the min_party method

* Fix the lint issues causing build failure

* Change the comment for test_min_party

* CR comments for modules/zk_concurrency.py

* Adding more examples for blocking/min_nodes

* Change the comment formatting

* Removing space at end of line

* Comments changes to rebuild

ezheidtmann committed:
Fix ImportError when using virtualbox cloud driver

Commit:
Merge pull request #32621 from ezheidtmann/virtualbox-driver-importerror

Fix ImportError when using virtualbox cloud driver

twangboy committed:
Fix mac_service and mac_system modules (#32587)

* Fix mac_service module

* Add integration tests for new functions

* Start will not enable the service beforehand

* Remove unused variables

aphor committed:
Issue32626 macports Portfile upgrade to 2015.8.8.2 (#32629)

* This is the 2015.8.5 Portfile distributed by MacPorts

* version to 2015.8.8.2 and use osx pkg LaunchDaemon.plist resources for master, api, syndic

Commit:
Add json5 renderer (#32635)

* Add JSON5 renderer

* Add documentation RST files for json5 renderer

nmadhok committed:
Fixing critical bug to remove only the specified Host instead of the entire Host cluster (#32641)

knuta committed:
Add support for unicode in all GECOS fields (#32630)

* Add support for unicode in all GECOS fields

* Stop exposing sdecode from salt.modules.useradd

* Slight rephrase of comment

* Fix lint errors

Commit:
Skipping boto test on OpenSuse (#32624)

tonybaloney committed:
Added global namespacing helpers to fix issue in dimensiondata module

tonybaloney commented:
I have generated a patch file for this to help out some internal testers, but should this be applied to develop or the 2016.3 branch? It was a bit of a silly mistake not to retest after removing the lines from PyLint's recommendation.

rallytime commented on salt/cloud/clouds/dimensiondata.py:
```diff
@@ -38,6 +38,7 @@
 
     HAS_LIBCLOUD = True
 except ImportError:
+    log.exception("You do not have apache-libcloud 0.20<= installed")
```
This warning needs to be removed as `log` isn't available at the point yet:
```
[ERROR   ] Failed to import clouds dimensiondata, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/root/SaltStack/salt/salt/loader.py", line 1296, in _load_module
    ), fn_, fpath, desc)
  File "/root/SaltStack/salt/salt/cloud/clouds/dimensiondata.py", line 41, in <module>
    log.exception("You do not have apache-libcloud 0.20<= installed")
NameError: name 'log' is not defined
```
The warning that `apache-libcloud` is missing should already be handled by the `get_dependencies` check in the `__virtual__()` function.

tonybaloney replied:
can https://github.com/saltstack/salt/blob/develop/salt/cloud/libcloudfuncs.py#L51 LIBCLOUD_MINIMAL_VERSION be specified per module?

rallytime replied:
Sure, but we should handle that in the module, rather than in that file. That's what the `__virtual__()` function can be used for. Similar to something like this: https://github.com/saltstack/salt/blob/develop/salt/modules/boto_lambda.py#L113

rallytime commented:
Adding this to 2016.3 is good as it is a needed bugfix and will be merged-forward to develop after this change is merged in.

tonybaloney committed:
Removed log warning message

Removed accidentally added git-am hangover added to WC

Removed duplicate lines

Removed trailing whitespace

rallytime commented:
@tonybaloney Can you clean up the pylint errors here? http://166.78.178.63:8080/job/PR/job/salt-pr-lint-n/1130/violations/file/salt/cloud/clouds/dimensiondata.py/

tonybaloney commented:
@rallytime fixed and rebased. I really need to write some tests on this module!

tonybaloney committed:
Removed trailing whitespace on comments

rallytime commented:
@tonybaloney It looks like you have some stray commits here from your rebase and we're unable to merge.

And tests are always welcome!

tonybaloney commented:
yeah that rebase screwed something up. I'm going to drop this PR and just apply the fix again in another one
